R/autocompletion.R

```diff
@@ -1,26 +1,98 @@
-#' @param activate bool default TRUE, enable chained auto-completion
-#' @name extra_auto_completion
-#' @return invisible NULL
-#' @noRd
+# Internal bookkeeping of the autocompletion mode ("rstudio" or "native")
+.polars_autocompletion = new.env(parent = emptyenv())
+
+#' Polars code completion
+#'
+#' @param mode One of `"auto"`, `"rstudio"`, or `"native"`. Automatic mode picks
+#' `"rstudio"` if `.Platform$GUI` is `"RStudio"`. `"native"` registers a custom
+#' line buffer completer with `utils:::rc.getOption("custom.completer")`.
+#' `"rstudio"` modifies RStudio code internal `.DollarNames` and function args
+#' completion, as the IDE does not behave well with
+#' `utils:::rc.getOption("custom.completer")`.
+#' @param verbose Print message of what mode is started.
 #'
-#' @details polars always supports auto completion via .DollarNames.
-#' However chained methods like x$a()$b()$? are not supported vi .DollarNames.
+#' @details
+#' Polars code completion has one implementation for a native terminal via
+#' `utils:::rc.getOption("custom.completer")` and one for Rstudio by intercepting
+#' Rstudio internal functions `.rs.getCompletionsFunction` &
+#' `.rs.getCompletionsDollar` in the loaded session environment `tools:rstudio`.
+#' Therefore, any error or slowness in the completion is likely to come from
+#' r-polars implementation.
 #'
-#' This feature experimental and not perfect. Any feedback is appreciated.
-#' Currently does not play that nice with Rstudio, as Rstudio backtick quotes any custom
-#' suggestions.
+#' Either completers will evaluate the full line-buffer to decide what methods
+#' are available. Pressing tab will literally evaluate left-hand-side with any
+#' following side. This works swiftly for the polars lazy API, but it can take
+#' some time for the eager API depending on the size of the data and of the
+#' query.
+#'
+#' @return NULL
+#' @export
 #'
 #' @examples
-#' # auto completion via .DollarNames method
-#' e = pl$lit(42) # to autocomplete pl$lit(42) save to variable
-#' # then write `e$`  and press tab to see available methods
+#' if (interactive()) {
+#'   # activate completion
+#'   polars_activate_code_completion()
+#'
+#'   # method / property completion for chained expressions
+#'   # add a $ and press tab to see methods of LazyFrame
+#'   pl$LazyFrame(iris)
 #'
-#' # polars has experimental auto completion for chain of methods if all on the same line
-#' pl$extra_auto_completion() # first activate feature (this will 'annoy' the Rstudio auto-completer)
-#' pl$lit(42)$to_series() # add a $ and press tab 1-3 times
-#' pl$extra_auto_completion(activate = FALSE) # deactivate
-pl_extra_auto_completion = function(activate = TRUE) {
+#'   # Arg + column-name completion
+#'   # press tab inside group_by() to see args and/or column names.
+#'   pl$LazyFrame(iris)$group_by()
+#'
+#'   # deactivate like this or restart R session
+#'   polars_deactivate_code_completion()
+#' }
+polars_activate_code_completion = function(
```

Comment:
I think a name like `polars_code_completion_*` is better than `polars_*_code_completion` in terms of autocompletion.

Reply:
Done